### DISCUSSION TITLE
`botorch` is loaded when `import baybe`

### DISCUSSION THREAD

**Discussion by rjavadi** [emdgroup/baybe#193] [Q&A]
Since `campaign` uses `recommender`s and they depend on `botorch`, we still relay on `torch`. If we lazy load `recommender`s in `campaign` like this:

```
if TYPE_CHECKING:
    from baybe.recommenders.base import RecommenderProtocol
    from baybe.recommenders.meta.sequential import TwoPhaseMetaRecommender
```
we can compeletly get rid of `torch`, but the tests fail with this error:

```
baybe/campaign.py:57: in Campaign
    recommender: RecommenderProtocol = field(factory=TwoPhaseMetaRecommender)
```
Now we should look for a workaround to use `TwoPhaseMetaRecommender` in `campaign`.

Here's the import tree of importing baybe:

![Screenshot 2024-04-03 at 12 46 37 PM](https://github.com/emdgroup/baybe/assets/5112312/ecebb12a-9b31-415e-9226-e268f691bc91)

And here's the import tree of baybe after lazy loading recommenders in campaign:

![Screenshot 2024-04-03 at 12 47 13 PM](https://github.com/emdgroup/baybe/assets/5112312/5e012bc3-0848-40a4-9de5-ac8572f17a3b)

**Comment by rjavadi** [comment]
Update: so far, I couldn't find a way to decouple or loosely  couple `Recommener` and `campaign`. As long as we import `recommender` in `baybe`, we load `torch` as well.

**Reply by AdrianSosic** [reply]
Hi @rjavadi, thanks for the update. I also had a brief look and know how to get rid of the remaining imports. However, it requires one additional change to happen first, which you should not have to bother with. Let me take care of the remaining open changes here so that you can fully focus on the polars part. Does that sound good to you?

**Reply by rjavadi** [reply]
Hi Adrian, thanks for the heads up.
Yes, sounds good.

**Comment by Scienfitz** [comment]
closing as no longer relevant
